hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/OzoneClientKeyReadWriteOps.java

```diff
@@ -0,0 +1,252 @@
+    for (int i = 0; i < clientsCount; i++) {
+      ozoneClients[i] = createOzoneClient(omServiceID, ozoneConfiguration);
+    }
+
+    ensureVolumeAndBucketExist(ozoneClients[0], volumeName, bucketName);
+    ozoneBuckets = new OzoneBucket[clientsCount];
+    for (int i = 0; i < clientsCount; i++) {
+      ozoneBuckets[i] = ozoneClients[i].getObjectStore().getVolume(volumeName)
+              .getBucket(bucketName);
+    }
+
+    timer = getMetrics().timer("key-read-write");
+    if (objectSizeInBytes >= 0) {
+      keyContent = RandomUtils.nextBytes(objectSizeInBytes);
+    }
+    if (kg == null) {
+      kg = new KeyGeneratorUtil();
+    }
+    runTests(this::readWriteKeys);
+
+    for (int i = 0; i < clientsCount; i++) {
+      if (ozoneClients[i] != null) {
+        ozoneClients[i].close();
+      }
+    }
+    return null;
+  }
+
+  public void readWriteKeys(long counter) throws Exception {
+    int clientIndex = (int)((counter) % clientsCount);
+    OzoneBucket ozoneBucket = ozoneBuckets[clientIndex];
+    TaskType taskType = decideReadOrWriteTask();
+    String keyName = getKeyName(clientIndex);
+
+    timer.time(() -> {
+      try {
+        switch (taskType) {
+        case READ_TASK:
+          processReadTasks(keyName, ozoneBucket);
+          break;
+        case WRITE_TASK:
+          processWriteTasks(keyName, ozoneBucket);
+          break;
+        default:
+          break;
+        }
+      } catch (RuntimeException ex) {
+        LOG.error(ex.getMessage());
+        throw ex;
+      } catch (IOException ex) {
+        LOG.error(ex.getMessage());
+        throw new RuntimeException(ex.getMessage());
+      }
+
+    });
+  }
+
+  public void processReadTasks(String keyName, OzoneBucket ozoneBucket)
+          throws RuntimeException, IOException {
+    if (readMetadataOnly) {
+      ozoneBucket.getKey(keyName);
+    } else {
+      byte[] data = new byte[objectSizeInBytes];
+      try (OzoneInputStream inputStream = ozoneBucket.readKey(keyName)) {
+        inputStream.read(data);
+      }
+    }
+  }
+  public void processWriteTasks(String keyName, OzoneBucket ozoneBucket)
+          throws RuntimeException, IOException {
+    try (OzoneOutputStream out =
+                 ozoneBucket.createKey(keyName, objectSizeInBytes)) {
+      out.write(keyContent);
+    } catch (Exception ex) {
+      throw ex;
+    }
+  }
+  public TaskType decideReadOrWriteTask() {
+    if (!isMixWorkload()) {
+      if (percentageRead == 100) {
+        return TaskType.READ_TASK;
+      } else {
+        return TaskType.WRITE_TASK;
+      }
+    }
+    //mix workload
+    int tmp = ThreadLocalRandom.current().nextInt(100) + 1; // 1 ~ 100
+    if (tmp < percentageRead) {
+      return TaskType.READ_TASK;
+    } else {
+      return TaskType.WRITE_TASK;
+    }
+  }
+
+  public String getKeyName(int clientIndex) {
+    int start, end;
+    // separate tasks evenly to each client
+    if (range < clientsCount) {
```

Comment:
Please validate the expected behavior when client >> range. It might make sense to cap the client to be <= range, so in the case client configured is less than range, we only create enough clients such that each client reads only one object. Or to hammer the same client from multiple clients, if range < clients, each client gets the entire range.

Reply:
ok I update the logic a little bit!
Each thread has its own client which holds tcp connection to OM.

We set the range here which would become the range of each thread/client could read/write on.

For example, If we set range = 10, start_index (index of key) = 20, threadNo. = 30, and we run the Freon test 5 times. 
Then only 5 clients would process the read/write operation from key index 20 to 29 (range=10).

If we set range = 10, start_index (index of key) = 20, threadNo. = 30 (clients = 30), and we run the Freon test 40 times. 
Then first 30 tests would be assign to 30 client to read/write from key index 20 to 29 (range=10). 
The rest of 10 tests would be picked up by any 10 clients whoever finish their tasks first.

Reply:
Then we utilize multiple nodes to separate the total range of keys the user want to test on.